File: polyplug.js
"use strict";
/******************************************************************************
PolyPlug 🦜🔌

A framework for linking the DOM and DOM based events with scripting languages
compiled to WASM. Part of the PyScript project.

See the README for more details, design decisions, and an explanation of how
things work.

This code does essentially five things:

1. Serialize and de-serialize nodes in the DOM via JSON.
2. Mutate nodes in the DOM to a new state.
3. Reach into the DOM to reference specific nodes.
4. Register and coordinate (remote) event handlers.
5. Use 1, 2, 3 and 4 while interacting, via message passing, with a scripting
   language.

Authors:
- Nicholas H.Tollervey (ntollervey@anaconda.org)

Copyright (c) 2022 Anaconda Inc. 

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
******************************************************************************/

const polyplug = function() {

    /**************************************************************************
    Internal state for PolyPlug.
    **************************************************************************/

    // Tracks event handler functions.
    const REGISTERED_EVENTS = {};

    /**************************************************************************
    Serialization functions for PolyPlug.
    **************************************************************************/

    function nodeToJS(node) {
        /*
        Takes a node in the DOM and recursively converts it and its children
        into a JSON encodable JavaScript object.

        What information is captured?

        * The nodeType (see: https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType).
        * The tag or node name (e.g. "div", "p" or "input").
        * The nodeValue (e.g. the textual content of a text node).
        * An object containing the key/value pairs representing the attributes
          of a node.
        * If a textarea, its "value" (i.e. textual content).
        * An array of childNodes, also captured in the same way. Textarea does
          not have childNodes.

        The nodeType is essential for representing the node with the correct
        "type" when de-serialized.
        */
        const obj = {
            nodeType: node.nodeType
        };
        if (node.tagName) {
            obj.tagName = node.tagName.toLowerCase();
        } else if (node.nodeName) {
            obj.nodeName = node.nodeName;
        }
        if (node.nodeValue) {
            obj.nodeValue = node.nodeValue;
        }
        const attrs = {};
        if (node.attributes && node.attributes.length > 0) {
            for (let i=0; i<node.attributes.length; i++) {
                attrs[node.attributes[i].nodeName] = node.attributes[i].nodeValue;
            }
        }
        if (Object.keys(attrs).length > 0) {
            obj.attributes = attrs;
        }
        if (obj.tagName === "textarea") {
            obj.value = node.value;
        }
        const childNodes = node.childNodes;
        const childNodesJS = [];
        // textarea nodes are special, in that they shouldn't have childNodes
        // since their content is supposed to be defined via their "value".
        if (obj.tagName !== 'textarea' && childNodes && childNodes.length > 0) {
            for (let i=0; i<childNodes.length; i++) {
                const child = nodeToJS(childNodes[i]);
                if (!(child.nodeValue && child.nodeValue.trim() === "")) {
                    childNodesJS.push(child);
                }
            }
        }
        obj.childNodes = childNodesJS;
        return obj;
    }

    function jsToNode(obj) {
        /*
        Takes a JavaScript object and returns the equivalent HTML node.

        * The nodeType indicates what sort of element to create with the
          document object.
        * The nodeValue indicates, for example, the textual content of a
          text node.
        * Attributes, if appropriate, are added to the node.
        * Child nodes are also recursively added, as appropriate.
        * A textarea is a bit special, see the comments in-line for more
          information.

        The resulting node should be merged into the DOM (rather than used to
        replace a node). This ensures event handlers are retained.
        */
        let node = null;  // The eventual result. 
        // See: https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType
        switch (obj.nodeType) {
            // ELEMENT_NODE
            case 1: {
                node = document.createElement(obj.tagName);
                if (obj.attributes && Object.keys(obj.attributes).length > 0) {
                    Object.entries(obj.attributes).map(attribute => {
                        node.setAttribute(attribute[0], attribute[1]);
                    })
                }
                if (obj.tagName === "textarea") {
                    node.value = obj.value;
                }
                break;
            }
            // TEXT_NODE
            case 3: {
                return document.createTextNode(obj.nodeValue);
            }
            // COMMENT_NODE
            case 8: {
                return document.createComment(obj.nodeValue);
            }
            // DOCUMENT_FRAGMENT_NODE
            case 11: {
                node = document.createDocumentFragment();
                break;
            }
            default: {
                // Default to an empty fragment node.
                return document.createDocumentFragment();
            }
        }
        // Recursively recreate the child nodes.
        // (textarea nodes are special, in that they shouldn't have childNodes
        // since their content is supposed to be defined via their "value".)
        if (obj.tagName !== 'textarea' && obj.childNodes && obj.childNodes.length > 0) {
            for (const childNode of obj.childNodes) {
                node.appendChild(jsToNode(childNode));
            }
        }
        return node;
    }

    /**************************************************************************
    DOM mutation functions for PolyPlug.
    **************************************************************************/

    function patch(oldNode, newNode) {
        /*
        Take an old HTML node and recursively patch its children with the new
        HTML node's children.

        Return a boolean indication if the patch resulted in any change to the
        old node.
        */
        let oldChild = oldNode.firstChild;  // current oldChild node.
        let newChild = newNode.firstChild;  // equivalent newChild node.
        let oldNextChild;  // to reference the next "old" node.
        let newNextChild;  // to reference the next equivalent "new" node.
        let changed = false;  // flag to indicate change to oldNode.

        // Handle pairs of equivalent old/new nodes.
        while (oldChild != null && newChild != null) {
            // Update references to the next nodes to handle after the
            // current ones.
            oldNextChild = oldChild.nextSibling;
            newNextChild = newChild.nextSibling;
            if(oldChild.nodeType !== newChild.nodeType || oldChild.nodeName !== newChild.nodeName) {
                // The old and new child nodes are completely different. Just
                // replace old with new.
                changed = true;
                oldNode.replaceChild(newChild, oldChild)
            } else if (oldChild.nodeType === oldChild.TEXT_NODE || oldChild.nodeType === oldChild.COMMENT_NODE) {
                // Check if textual content of TEXT_NODEs or COMMENT_NODEs
                // needs updating.
                if (oldChild.nodeValue !== newChild.nodeValue) {
                    changed = true;
                    oldChild.nodeValue = newChild.nodeValue;
                }
            } else {
                // In all other cases, patch the attributes and recurse into
                // the branches of the DOM tree.
                changed = patchAttrs(oldChild, newChild) || changed;
                changed = patch(oldChild, newChild) || changed;
            }
            // Swap to next nodes. Rinse and repeat. ;-)
            oldChild = oldNextChild;
            newChild = newNextChild;
        }
        if (newChild) {
            // Handle a "spare" newChild node that doesn't have an oldChild
            // equivalent to patch.
            do {
                // Just keep adding the spare newChild nodes to the oldNode
                // parent.
                changed = true;
                newNextChild = newChild.nextSibling;
                oldNode.appendChild(newChild);
                // Swap, rinse, repeat.
                newChild = newNextChild;
            } while (newChild);
        }
        if (oldChild) {
            // Handle a "spare" oldChild node that doesn't have a newChild
            // equivalent. I.e. the oldChild is to be removed.
            do {
                // Just keep removing the spare oldChild nodes from the oldNode
                // parent.
                changed = true;
                oldNextChild = oldChild.nextSibling;
                oldNode.removeChild(oldChild)
                // Swap, rinse, repeat.
                oldChild = oldNextChild;
            } while (oldChild);
        }
        // Return if the oldNode has changed as a result of the patch.
        return changed;
    }

    function patchAttrs(oldNode, newNode) {
        /*
        Takes an old node and updates its attributes to those from the new
        node.

        Return a boolean indication if any of the attributes have actually
        changed as a result of the patch.
        */
        let changed = false;

        // Check for changes in attributes that already exist for the old node.
        let i = 0;
        if (oldNode.attributes) {
            i = oldNode.attributes.length;
            while (i--) {
                const oldAttr = oldNode.attributes[i];
                const newAttr = newNode.attributes[oldAttr.name];
                if (newAttr == null) {
                    // Remove the attribute, it doesn't exist on the newNode.
                    changed = true;
                    oldNode.removeAttribute(oldAttr.name);
                } else if (oldAttr.nodeValue !== newAttr.nodeValue) {
                    // Attribute's value needs updating to newNode's.
                    changed = true;
                    oldAttr.nodeValue = newAttr.nodeValue;
                }
            }
        }
        // Check for new attributes to add to the old node.
        if (newNode.attributes) {
            i = newNode.attributes.length;
            while (i--) {
                const {name, nodeValue} = newNode.attributes[i];
                // Ignore attributes that already exist on oldNode.
                if (!oldNode.hasAttribute(name)) {
                    changed = true;
                    oldNode.setAttribute(name, nodeValue);
                }
            }
        }
        // Return if the oldNode's attributes have actually changed as a
        // result of the patch.
        return changed;
    }

    function mutate(oldNode, newNode) {
        /*
        An unsophisticated DOM mutator.

        Take an old HTML node and recursively mutate it to the new HTML node's
        configuration.

        Return a boolean indication to show if the mutation resulted in any
        change to the old node.
        */
        const changed = patchAttrs(oldNode, newNode)
        return patch(oldNode, newNode) || changed;
    }

    /**************************************************************************
    DOM reference functions for PolyPlug.
    **************************************************************************/

    function getElements(q) {
        /*
        Return a collection of HTML element[s] matching the query object "q".

        In order of precedence, the function will try to find results by id or
        tag or classname or CSS selector (but not a combination thereof).

        If no matches found, returns null.
        */
        if (q.id) {
            // getElementById doesn't return a collection, hence wrapping the
            // result as an array, for the sake of consistency.
            const result = [];
            const element = document.getElementById(q.id);
            if (element) {
                result.push(element);
            }
            return result;
        } else if (q.tag) {
            return document.getElementsByTagName(q.tag);
        } else if (q.classname) {
            return document.getElementsByClassName(q.classname);
        } else if (q.css) {
            return document.querySelectorAll(q.css);
        }
        return null;  // Explicit > implicit. ;-)
    }

    /**************************************************************************
    Event handling functions for PolyPlug.
    **************************************************************************/

    function registerEvent(query, eventType, listener) {
        /*
        Register an event listener, given:

        * target element[s] via a query object (see getElements),
        * the event type (e.g. "click"), and,
        * the name of the listener to call in the remote interpreter.

        The handler function is called by dispatching a polyplugSend event
        with details of the event encoded as a JSON string.

        Depending upon how the remote interpreter is run (on the main thread,
        in a web worker, etc), this event should be handled to call the
        expected function in the most appropriate way.
        */
        const elements = getElements(query);
        const eventHandler = function(e) {
            const detail = JSON.stringify({
                type: e.type,
                target: nodeToJS(e.target),
                listener: listener
            });
            const send = new CustomEvent("polyplugSend", {detail: detail});
            document.dispatchEvent(send);
        }
        REGISTERED_EVENTS[listener] = eventHandler;
        elements.forEach(function(element) {
            element.addEventListener(eventType, eventHandler);
        });
    }

    function removeEvent(query, eventType, listener) {
        /*
        Remove an event listener, given:

        * target element[s] via a query object (see getElements),
        * the event type (e.g. "click"), and,
        * the name of the listener to call in the remote interpreter.
        */
        const elements = getElements(query);
        const eventHandler = REGISTERED_EVENTS[listener];
        if (eventHandler) {
            elements.forEach(function(element) {
                element.removeEventListener(eventType, eventHandler);
            });
            delete REGISTERED_EVENTS[listener];
        }
    }

    /**************************************************************************
    Message handling functions for PolyPlug.
    **************************************************************************/

    function receiveMessage(raw) {
        /*
        Receive a raw message string (containing JSON). Deserialize it and
        dispatch the message to the appropriate handler function.
        */
        try {
            const message = JSON.parse(raw);
            switch (message.type) {
                case "updateDOM":
                    onUpdateDOM(message);
                    break;
                case "registerEvent":
                    onRegisterEvent(message);
                    break;
                case "removeEvent":
                    onRemoveEvent(message);
                    break
                case "stdout":
                    onStdout(message);
                    break;
                case "stderr":
                    onStderr(message);
                    break;
                case "error":
                    onError(message);
                    break;
                default:
                    console.log("Unknown message type.")
                    console.log(message)
                    break;
            }
        } catch (error) {
            onError(error);
        }
    }

    function onUpdateDOM(msg) {
        /*
        Handle messages from an interpreter to update the DOM.

        Given a query to identify a target element, mutate it to the target
        state (encoded as JSON).

        Sample message:

        msg = {
            type: "updateDOM",
            query: {
              id: "idOfDomElementToMutate"
            },
            target: {
                ... JSON representation of the target state ...
            }
        }
        */
        const elements = getElements(msg.query);
        if (elements.length === 1) {
            // Update the single valid match to 
            const oldElement = elements[0];
            const newElement = jsToNode(msg.target)
            mutate(oldElement, newElement);
        }
    }

    function onRegisterEvent(msg) {
        /*
        Handle requests from an interpreter to register an event listener.

        Given a query to identify target element[s], listen for the event type
        and handle with the referenced listener function in the remote
        interpreter.

        Sample message:

        msg = {
            type: "registerEvent",
            query: {
                id: "idOfDomElement"
            },
            eventType: "click",
            listener: "my_on_click_function"
        }
        */
        registerEvent(msg.query, msg.eventType, msg.listener);
    }

    function onRemoveEvent(msg) {
        /*
        Handle requests to unbind the referenced event type on the elements
        matched by the query.

        Sample message:

        msg = {
            type: "removeEvent",
            query: {
                id: "idOfDomElement"
            },
            eventType: "click",
        }
        */
        removeEvent(msg.query, msg.eventType);
    }

    function onStdout(msg) {
        /*
        Handle "normal" STDOUT output.

        Simply dispatch a "polyplugStdout" custom event with the event's detail
        containing the emitted characters.

        Sample message:

        msg = {
            type: "stdout",
            content: "Some stuff to print to the terminal"
        }
        */
        const polyplugStdout = new CustomEvent("polyplugStdout", {detail: msg.content});
        document.dispatchEvent(polyplugStdout);
    }

    function onStderr(msg) {
        /*
        Handle "normal" STDERR output.

        Simply dispatch a "polyplugStderr" custom event with the event's detail
        containing the emitted characters.

        Sample message:

        msg = {
            type: "stderr",
            content: "Some stuff from stderr"
        }
        */
        const polyplugStderr = new CustomEvent("polyplugStderr", {detail: msg.content});
        document.dispatchEvent(polyplugStderr);
    }

    function onError(msg) {
        /*
        Handle generic error conditions from the interpreter.

        Dispatch a "polyplugError" custom event with the event detail set to
        the error context.

        Sample message:

        msg = {
            type: "error",
            context: {
                ... arbitrary data about the error condition ...
                (e.g. Exception name, line number, stack trace etc)
            }
        }
        */
        const polyplugError = new CustomEvent("polyplugError", {detail: msg.context});
        document.dispatchEvent(polyplugError);
    }

    return {
        nodeToJS: nodeToJS,
        jsToNode: jsToNode,
        mutate: mutate,
        getElements: getElements,
        registerEvent: registerEvent,
        removeEvent: removeEvent,
        receiveMessage: receiveMessage
    }
};

File: polyplug.py
import binascii
import builtins
import hashlib
import json


# Registered event listeners.
LISTENERS = {}


class Query:
    """
    Encapsulates a query to match elements in the DOM.

    A query must have one, and only one, of:

    * id - a unique element id (e.g. "myElementId").
    * classname - a class (e.g. "my-css-class").
    * tag - a tag name (e.g. "p").
    * css - a css selector (e.g "p.className").

    These types of query relate to the four ways in which elements in an HTML
    document may be identified.

    A Query is instantiated with a string using the following common syntax
    for identifying the different types of query:

    * #id-name - Starting with a hash indicates an element's unique id.
    * .class-name - Starting with a full stop indicates a CSS class name.
    * p - A string of only alphabetical characters is treated as a tag name.
    * p.classname - Anything else is assumed to be a css selector.

    E.g.

    q = Query(".myElementID")
    """

    def __init__(self, query):
        """
        Validate and interpret the query string.
        """
        self.raw_query = query
        target = None
        if query[:1] == "#":
            # Unique id.
            target = query[1:]
            if target:
                self.query_type = "id"
            else:
                raise ValueError("Invalid id.")
        elif query[:1] == ".":
            # CSS class.
            target = query[1:]
            if target:
                self.query_type = "classname"
            else:
                raise ValueError("Invalid class.")
        elif query.isalpha():
            # tagName.
            target = query
            self.query_type = "tag"
        else:
            # CSS selector.
            target = query
            self.query_type = "css"

        if target:
            setattr(self, self.query_type, target)
        else:
            raise ValueError("Bad query specification.")

    @property
    def as_dict(self):
        """
        Return a dict to be JSON encoded.
        """
        return {
            self.query_type: getattr(self, self.query_type),
        }


class DomEvent:
    """
    Represents an event dispatched in the DOM.

    The event_type (e.g. "click") and target (DOM element that was the source
    of the event) are referenced.
    """

    def __init__(self, event_type, target):
        self.event_type = event_type
        self.target = target


class HTMLTokenizer:
    """
    Turns a string into a structured representation of the DOM.

    MUST USE XHTML (i.e. open and closing tags).
    """

    QUOTES = "\"'"

    def __init__(self, raw):
        self.raw = raw
        self.len = len(raw)
        self.pos = 0
        self.char = self.next_char()

    def next_char(self):
        """
        Get the next character, or return an empty string for EOF or no raw
        input.
        """
        if self.len and self.pos < self.len:
            self.char = self.raw[self.pos]
            self.pos += 1
            return self.char
        else:
            self.char = ""
        return ""

    def get_char(self):
        """
        Get the current character, and step to the next one.
        """
        result = self.char
        self.next_char()
        return result

    def skip_ws(self):
        """
        Skip over whitespace.
        """
        while self.char.isspace():
            self.next_char()

    def match(self, expected):
        """
        Return a boolean indication if the next character[s] is the expected
        character.

        Expected could be a multi-character match (mainly used for single and
        double quotes). E.g. expected = "\"'"
        """
        self.skip_ws()
        if self.char in expected:
            self.next_char()
            return True
        return False

    def expect(self, expected):
        """
        Raise an exception if the expected character[s] is not matched.
        """
        if not self.match(expected):
            raise ValueError("Bad HTML syntax.")

    def get_name(self):
        """
        Get the name of a tag or attribute.

        E.g. used to extract "div" and "id" from this fragment:

        <div id="foo">
        """
        self.skip_ws()
        result = ""
        while True:
            c = self.char
            if not (c.isalpha() or c.isdigit() or c in "_-."):
                break
            result += self.get_char()
        return result

    def get_value(self):
        """
        Get the value associated with an attribute.

        E.g. used to extract the "foo" value (without quotes) from this
        fragment:

        <div id="foo">
        """
        self.skip_ws()
        result = ""
        try:
            self.expect("=")
            self.expect(self.QUOTES)
            while True:
                c = self.char
                if not (c.isalpha() or c.isdigit() or c in "_-. ;:!"):
                    break
                result += self.get_char()
            self.expect(self.QUOTES)
            return result
        except ValueError:
            return ""

    def get_attrs(self):
        """
        Return an Attributes instance representing any attributes attached to
        an ElementTag.
        """
        attrs = Attributes()
        name = self.get_name()
        while name:
            value = self.get_value()
            attrs[name] = value
            name = self.get_name()
        return attrs

    def get_text(self, until="<"):
        """
        Return textual content until the start of a new Node ("<") or until
        "until" matches.
        """
        result = ""
        until_len = len(until)
        while result[-until_len:] != until and self.char:
            result += self.get_char()
        if result[-until_len:] == until:
            self.pos = self.pos - until_len
            if self.char:
                self.pos = self.pos - 1
            self.next_char()
            return result[:-until_len]
        else:
            # EOF
            return result

    def tokenize(self, parent=None):
        """
        Tokenize the raw HTML input and return a DOM representation using the
        Node and Attributes classes defined below.

        The parent ElementNode is given since we're always parsing its
        innerHTML.
        """
        if not isinstance(parent, ElementNode):
            raise ValueError("Parent must be an ElementNode")
        current_node = None
        current_parent = parent
        while self.char:
            if self.match("<"):
                # Tag opens.
                if self.match("/"):
                    # End tag. Close and check depth of tree.
                    # Get the name of the closing tag.
                    name = self.get_name()
                    if current_node and name == current_node.tagName:
                        # Close current node and continue at current depth.
                        current_parent.add_child(current_node)
                        current_node = None
                    elif name == current_parent.tagName:
                        # Step back up the tree to the parent context.
                        current_node = current_parent
                        current_parent = current_node.parent
                        current_parent.add_child(current_node)
                        current_node = None
                    else:
                        # Unexpected close tag.
                        raise ValueError("Unexpected close tag.", name)
                    self.expect(">")
                elif self.match("?"):
                    # XML prolog - consume and ignore.
                    self.get_attrs()
                    self.expect("?")
                    self.expect(">")
                elif self.match("!"):
                    # CommentNode - get nodeValue.
                    self.expect("-")
                    self.expect("-")
                    value = ""
                    while True:
                        value += self.get_char()
                        if value[-3:] == "-->":
                            break
                    comment = CommentNode(nodeValue=value[:-3])
                    current_parent.add_child(comment)
                else:
                    # ElementNode
                    tagName = self.get_name()
                    attrs = self.get_attrs()
                    if current_node:
                        current_parent = current_node
                    if tagName == "textarea":
                        self.expect(">")
                        value = self.get_text(until="</textarea>")
                        textarea_node = ElementNode(
                            tagName=tagName,
                            attributes=attrs,
                            value=value,
                            parent=current_parent,
                        )
                        for c in "</textarea>":
                            self.expect(c)
                        current_parent.add_child(textarea_node)
                    else:
                        current_node = ElementNode(
                            tagName=tagName,
                            attributes=attrs,
                            parent=current_parent,
                        )
                        self.expect(">")
            else:
                # TextNode
                value = self.get_text()
                text = TextNode(nodeValue=value)
                if current_node:
                    current_node.add_child(text)
                else:
                    current_parent.add_child(text)


class Node:
    """
    Represents a node in the DOM.
    """

    def __init__(self, **kwargs):
        self._node = kwargs
        self.parent = kwargs.get("parent")

    @property
    def outerHTML(self):
        """
        Get a string representation of the element's outer HTML.
        """
        return NotImplemented

    @property
    def as_dict(self):
        """
        JSON serializable.
        """
        return NotImplemented


class Attributes(dict):
    """
    Represents a collection of attributes attached to an ElementNode.
    """

    def __getattr__(self, key):
        if key in self:
            return self[key]
        else:
            raise AttributeError("No such attribute: " + key)

    def __setattr__(self, key, value):
        self[key] = value

    def __delattr__(self, key):
        if key in self:
            del self[key]
        else:
            raise AttributeError("No such attribute: " + key)


class ElementNode(Node):
    """
    An element defined by a tag, may have attributes and children.
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.tagName = kwargs["tagName"]
        self.attributes = Attributes(kwargs.get("attributes", {}))
        if self.tagName == "textarea":
            # The textarea doesn't have children. Only a text value.
            self.value = kwargs.get("value", "")

    def add_child(self, child):
        """
        Add a child node to the children of this node.
        """
        children = self._node.get("childNodes", [])
        node_dict = child
        if isinstance(child, Node):
            node_dict = child.as_dict
        node_dict["parent"] = self
        children.append(node_dict)
        self._node["childNodes"] = children

    @property
    def outerHTML(self):
        """
        Get a string representation of the element's outer HTML.
        """
        result = "<" + self.tagName
        for attr, val in self.attributes.items():
            result += " " + attr + '="' + val + '"'
        result += ">"
        if self.tagName == "textarea":
            result += self.value
        else:
            result += self.innerHTML
        result += "</" + self.tagName + ">"
        return result

    @property
    def innerHTML(self):
        """
        Get a string representation of the element's inner HTML.
        """
        result = ""
        for child in self.childNodes:
            result += child.outerHTML
        return result

    @innerHTML.setter
    def innerHTML(self, raw):
        """
        Use the raw innerHTML to create children.
        """
        self._node["childNodes"] = []
        tok = HTMLTokenizer(raw)
        tok.tokenize(self)

    @property
    def childNodes(self):
        """
        A data structure representing the tree of child nodes.
        """
        if self.tagName == "textarea":
            # The textarea doesn't have children. Only a text value.
            return []
        else:
            children = self._node.get("childNodes", [])
            result = []
            for child in children:
                node_type = child.get("nodeType")
                child["parent"] = self
                if node_type == 1:
                    # ELEMENT_NODE
                    result.append(ElementNode(**child))
                elif node_type == 3:
                    # TEXT_NODE
                    result.append(TextNode(**child))
                elif node_type == 8:
                    # COMMENT_NODE
                    result.append(CommentNode(**child))
                else:
                    # FRAGMENT_NODE
                    result.append(FragmentNode(**child))
            return result

    @property
    def as_dict(self):
        """
        JSON serializable.
        """
        result = {
            "nodeType": 1,
            "tagName": self.tagName,
            "childNodes": [child.as_dict for child in self.childNodes],
        }
        if self.attributes:
            result["attributes"] = self.attributes
        if self.tagName == "textarea":
            result["value"] = self.value
        return result

    def find(self, selector):
        """
        Recursively check this node, and its children, and return a result
        representing nodes that match "selector" string. The selector
        string understands:

        * #my-id - returns the first ElementNode with the id "my-id". Returns
          None if no match is found.
        * .my-class - returns a list containing elements with the class
          "my-class". Returns an empty list if no match is found.
        * li - returns a list containing elements with the tagName "li".
          Returns an empty list if no match is found.
        """
        # Validate selector.
        if not selector:
            raise ValueError("Missing selector.")
        if selector[0] == "#":
            # Select by unique id.
            target_id = selector[1:]
            if target_id:
                return self._find_by_id(target_id)
            else:
                raise ValueError("Invalid id.")
        elif selector[0] == ".":
            # Select by css class.
            target_class = selector[1:]
            if target_class:
                return self._find_by_class(target_class)
            else:
                raise ValueError("Invalid class.")
        else:
            # select by tagName.
            if selector.isalpha():
                return self._find_by_tagName(selector.lower())
            else:
                raise ValueError("Invalid tag name.")

    def _find_by_id(self, target):
        """
        Return this node, or the first of its children, if it has the id
        attribute of target. Returns None if no match is found.
        """
        my_id = self.attributes.get("id")
        if my_id and my_id == target:
            return self
        else:
            for child in (
                node
                for node in self.childNodes
                if isinstance(node, ElementNode)
            ):
                result = child._find_by_id(target)
                if result:
                    return result
        return None

    def _find_by_class(self, target):
        """
        Return a list containing this node, or any of its children, if the
        node has the associated target class.
        """
        result = []
        class_attr = self.attributes.get("class", "")
        if class_attr:
            classes = [
                class_name
                for class_name in class_attr.split(" ")
                if class_name
            ]
            if target in classes:
                result.append(self)
        for child in (
            node for node in self.childNodes if isinstance(node, ElementNode)
        ):
            result.extend(child._find_by_class(target))
        return result

    def _find_by_tagName(self, target):
        """
        Return a list containing this node, or any of its children, if the
        node has the associated target as its tagName (e.g. "div" or "p").
        """
        result = []
        if self.tagName == target:
            result.append(self)
        for child in (
            node for node in self.childNodes if isinstance(node, ElementNode)
        ):
            result.extend(child._find_by_tagName(target))
        return result


class TextNode(Node):
    """
    Textual content inside an ElementNode.
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.nodeValue = kwargs.get("nodeValue")

    @property
    def outerHTML(self):
        """
        Get a string representation of the element's outer HTML.
        """
        return self.nodeValue

    @property
    def as_dict(self):
        """
        JSON serializable.
        """
        return {
            "nodeType": 3,
            "nodeName": "#text",
            "nodeValue": self.nodeValue,
            "childNodes": [],
        }


class CommentNode(Node):
    """
    A comment such as <!-- ... -->
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.nodeValue = kwargs.get("nodeValue")

    @property
    def outerHTML(self):
        """
        Get a string representation of the element's outer HTML.
        """
        return "<!--" + self.nodeValue + "-->"

    @property
    def as_dict(self):
        """
        JSON serializable.
        """
        return {
            "nodeType": 8,
            "nodeName": "#comment",
            "nodeValue": self.nodeValue,
            "childNodes": [],
        }


class FragmentNode(Node):
    """
    A minimal node when no other type is appropriate.
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    @property
    def outerHTML(self):
        """
        Get a string representation of the element's outer HTML.
        """
        return ""

    @property
    def as_dict(self):
        """
        JSON serializable.
        """
        return {"nodeType": 11, "childNodes": []}


def get_listener_id(query, event_type, listener):
    """
    Given a query, event type and listener function, generate a unique id from
    this combination.
    """
    raw = query.raw_query + event_type + listener.__name__
    return binascii.hexlify(
        hashlib.sha256(raw.encode("utf-8")).digest()
    ).decode("ascii")


def print(*args, **kwargs):
    """
    Overridden print so output is handled correctly via JSON message passing
    instead of just raw text.
    """
    sep = kwargs.get("sep", " ")
    end = kwargs.get("end", "\n")
    content = sep.join(args) + end
    builtins.print(json.dumps({"type": "stdout", "content": content}))


def update(query, target):
    """
    Update the DOM so the node[s] matching the query are mutated to the state
    defined by the target node.
    """
    q = Query(query)
    builtins.print(
        json.dumps(
            {
                "type": "updateDOM",
                "query": q.as_dict,
                "target": target.as_dict,
            }
        )
    )


def remove(query, event_type, listener):
    """
    Remove the referenced listener from handling the event_type from the
    node[s] matching the query.
    """
    q = Query(query)
    listener_id = get_listener_id(q, event_type, listener)
    del LISTENERS[listener_id]
    builtins.print(
        json.dumps(
            {
                "type": "removeEvent",
                "query": q.as_dict,
                "eventType": event_type,
            }
        )
    )


def plug(query, event_type):
    """
    Plug a Python function into a DOM event.

    The event is specified by a query to match elements in the DOM tree, and
    an event_type (e.g. "click").

    Returns a decorator function that wraps the user's own function that is
    to be registered.

    This decorator wrapper function creates a closure in which various
    contextual aspects are contained and run.
    """
    q = Query(query)

    def decorator(fn):
        """
        Register the function via the query and event_type.
        """

        def wrapper(event):
            return fn(event)

        listener_id = get_listener_id(q, event_type, wrapper)
        builtins.print(
            json.dumps(
                {
                    "type": "registerEvent",
                    "query": q.as_dict,
                    "eventType": event_type,
                    "listener": listener_id,
                }
            )
        )

        LISTENERS[listener_id] = wrapper
        return wrapper

    return decorator


def receive(raw):
    """
    Given a raw JSON message, decode it, find the expected handler function,
    re-constitute the DOM, and call the handler with the appropriate context.
    """
    try:
        msg = json.loads(raw)
        event_type = msg.get("type")
        target = msg.get("target")
        listener = msg.get("listener")
        if event_type and target and listener:
            if listener in LISTENERS:
                event = DomEvent(event_type, ElementNode(**target))
                LISTENERS[listener](event)
            else:
                raise RuntimeError("No such listener: " + listener)
        else:
            raise ValueError("Incomplete message received: " + raw)
    except Exception as ex:
        context = {"type": type(ex).__name__, "msg": str(ex)}
        builtins.print(
            json.dumps(
                {
                    "type": "error",
                    "context": context,
                }
            )
        )
